client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -280,7 +280,7 @@ private static List<Column> getColumns(StatementClient client, Consumer<QuerySta
         verify(client.isFinished());
         QueryStatusInfo results = client.finalStatusInfo();
         if (results.getError() == null) {
-            throw new SQLException(format("Query has no columns (#%s)", results.getId()));
+            return ImmutableList.of();
```

Comment:
I am not sure this is a good change.

It looks we're going to call `client.advance();` until we drain the results.
I don't know what are the assumptions for the code above (it originates a while back from 8ccea3a813bc4a00b3969e6cbcb40601f655f1b5 and doesn't carry much comments), but it feels wrong to me.

I think the intention here is

- call `client.advance()` until we know the columns of the result, since this information is one of the first send back by the server

if the server now doesn't return columns in some cases, maybe it needs to indicate no columns will ever be provided, so that clients don't need to drain / buffer all the results?

eg some related code in the python client
https://github.com/trinodb/trino-python-client/blob/cffd2b232d9298321c34430f0ac829f6b6ebea8f/trino/client.py#L652-L655

cc. @electrum

Reply:
> if the server now doesn't return columns in some cases, maybe it needs to indicate no columns will ever be provided, so that clients don't need to drain / buffer all the results?

It always returns the columns, eventually, unless the query fails before it gets analyzed. But it may not happen in the first (or first few) requests, depending on how long the query gets queued, etc.

Reply:
> But it may not happen in the first (or first few) requests

that's true

> It always returns the columns, eventually

@martint  Do we always need to produce columns, also for `INSERT`, `CALL`, `SET` or `GRANT` statements?